File: github_stats.py
import requests
import json
from collections import Counter

USERNAME = "girum54"

def fetch_github_stats():
    """Fetch GitHub stats for the user"""
    headers = {"Accept": "application/vnd.github.v3+json"}
    
    # Get user profile
    user_url = f"https://api.github.com/users/{USERNAME}"
    user_response = requests.get(user_url, headers=headers)
    user_data = user_response.json()
    
    # Get repositories
    repos_url = f"https://api.github.com/users/{USERNAME}/repos?per_page=100"
    repos_response = requests.get(repos_url, headers=headers)
    repos_data = repos_response.json()
    
    return user_data, repos_data

def calculate_language_stats(repos):
    """Calculate language statistics from repositories"""
    languages = []
    for repo in repos:
        if repo.get("language"):
            languages.append(repo["language"])
    
    lang_counts = Counter(languages)
    total = sum(lang_counts.values())
    
    lang_stats = []
    for lang, count in lang_counts.most_common(10):
        percentage = (count / total) * 100 if total > 0 else 0
        lang_stats.append({
            "language": lang,
            "count": count,
            "percentage": round(percentage, 1)
        })
    
    return lang_stats

def generate_markdown(user_data, lang_stats):
    """Generate markdown content for README"""
    md = f"""# Hi! Girum Tilahun here.

<br>

## Languages

| Language | Repos | % |
|----------|-------|---|
"""
    for stat in lang_stats:
        md += f"| {stat['language']} | {stat['count']} | {stat['percentage']}% |\n"
    
    md += """
<br><br>

## ✨ Stats

"""
    md += f"- **Public Repos**: {user_data.get('public_repos', 0)}\n"
    md += f"- **Followers**: {user_data.get('followers', 0)}\n"
    md += f"- **Following**: {user_data.get('following', 0)}\n"
    md += f"- **Account Created**: {user_data.get('created_at', 'N/A')[:10]}\n"
    
    md += """
<br><br>

```
💌 girumtilahun54@gmail.com
```
"""
    return md

def main():
    print(f"Fetching stats for {USERNAME}...")
    user_data, repos_data = fetch_github_stats()
    
    if "message" in user_data:
        print(f"Error: {user_data['message']}")
        return
    
    lang_stats = calculate_language_stats(repos_data)
    markdown = generate_markdown(user_data, lang_stats)
    
    # Write to README
    with open("README.md", "w", encoding="utf-8") as f:
        f.write(markdown)
    
    print("✅ README.md updated successfully!")
    print(f"\nStats summary:")
    print(f"- Total repos: {len(repos_data)}")
    print(f"- Languages: {len(lang_stats)}")

if __name__ == "__main__":
    main()

File: github_stats.js
const https = require('https');

const USERNAME = 'girum54';

function fetch(url) {
    return new Promise((resolve, reject) => {
        const options = {
            headers: {
                'User-Agent': 'GitHub-Stats-Script'
            }
        };
        https.get(url, options, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    resolve(JSON.parse(data));
                } catch (e) {
                    reject(new Error(`Failed to parse JSON: ${data}`));
                }
            });
        }).on('error', reject);
    });
}

async function fetchGitHubStats() {
    const userUrl = `https://api.github.com/users/${USERNAME}`;
    const reposUrl = `https://api.github.com/users/${USERNAME}/repos?per_page=100`;
    
    const [userData, reposData] = await Promise.all([
        fetch(userUrl),
        fetch(reposUrl)
    ]);
    
    return { userData, reposData };
}

function calculateLanguageStats(repos) {
    const languages = [];
    repos.forEach(repo => {
        if (repo.language) {
            languages.push(repo.language);
        }
    });
    
    const langCounts = {};
    languages.forEach(lang => {
        langCounts[lang] = (langCounts[lang] || 0) + 1;
    });
    
    const total = languages.length;
    const langStats = Object.entries(langCounts)
        .map(([lang, count]) => ({
            language: lang,
            count,
            percentage: total > 0 ? ((count / total) * 100).toFixed(1) : 0
        }))
        .sort((a, b) => b.count - a.count)
        .slice(0, 10);
    
    return langStats;
}

function generateMarkdown(userData, langStats) {
    let md = `# Hi! Girum Tilahun here.

<br>

## Languages

| Language | Repos | % |
|----------|-------|---|
`;
    langStats.forEach(stat => {
        md += `| ${stat.language} | ${stat.count} | ${stat.percentage}% |\n`;
    });
    
    md += `
<br><br>

## ✨ Stats

`;
    md += `- **Public Repos**: ${userData.public_repos || 0}\n`;
    md += `- **Followers**: ${userData.followers || 0}\n`;
    md += `- **Following**: ${userData.following || 0}\n`;
    md += `- **Account Created**: ${userData.created_at ? userData.created_at.split('T')[0] : 'N/A'}\n`;
    
    md += `
<br><br>

\`\`\`
💌 girumtilahun54@gmail.com
\`\`\`
`;
    return md;
}

async function main() {
    console.log(`Fetching stats for ${USERNAME}...`);
    
    try {
        const { userData, reposData } = await fetchGitHubStats();
        
        if (userData.message) {
            console.error(`Error: ${userData.message}`);
            return;
        }
        
        const langStats = calculateLanguageStats(reposData);
        const markdown = generateMarkdown(userData, langStats);
        
        const fs = require('fs');
        fs.writeFileSync('README.md', markdown, 'utf-8');
        
        console.log('✅ README.md updated successfully!');
        console.log(`\nStats summary:`);
        console.log(`- Total repos: ${reposData.length}`);
        console.log(`- Languages: ${langStats.length}`);
    } catch (error) {
        console.error('Error:', error.message);
    }
}

main();
